spec/function.dd

```diff
@@ -173,124 +173,317 @@ int foo();
-    references of the result may be assumed to not refer to any object that
-    existed before the function call. For example:)
+$(LI The function is not a constructor or a destructor.)
+)
+
+$(H3 Weak Purity)
+
+$(P A function annotated with `pure` is weakly pure if and only if it passes the `pure` checks
+but not the strong purity checks. Explicitly, weakly pure meet $(I at least one) of the
+following conditions:)
+
+$(OL
+$(LI One or more parameter types are incompletely defined, transitively.)
+
+$(LI One or more parameters allow indirect access$(MDASH)through `ref` parameters and/or by
+transitively following class references, dynamic arrays, or pointers$(MDASH)to unqualified data or
+data with qualifiers other than `immutable`. Example:)
+
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+// These functions are weakly pure:
+pure int f1(Object obj);
+pure int f2(int* x);
+pure int f3(const int[] x);
+struct S
+{
+    const int x;
+}
+pure int f4(S* object);
+---
+)
+
+$(LI The returned type allows indirect access$(MDASH)through `ref` return and/or by transitively
+following class references, dynamic arrays, or pointers$(MDASH)to unqualified data or data with
+qualifiers other than `immutable`. Example:)
+
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+// These functions are weakly pure:
+pure Object f1(int);
+pure int* f2(int);
+pure const(int[]) f3(int);
+struct S
+{
+    const int x;
+}
+pure S* f4(int);
+---
+)
+
+$(LI The function is a constructor or destructor.)
+)
+
+$(H3 Strongly Pure Functions and Optimizations)
+
+$(P Strong purity enables optimizations based on their inability to read or modify data outside
+their parameters. In turn, that guarantees the function always returns the same result for the
+same arguments.)
+
+$(P A strongly pure function that accepts and returns by value can be assumed to return the same
+value for the same inputs. The compiler is allowed to memoize the result and reuse it for
+subsequent calls with identical arguments. Example:)
+
+---
+pure double fun(double);
+auto x = fun(1) + fun(1);
+// Code generated may be:
+// auto __t = fun(1);
+// auto x = __t + __t;
+---
+
+$(P For the purpose of deeming two calls to a strongly pure function equivalent, all arguments must
+be bitwise equal (as per the `is` operator), not logically equal (as per the `==` operator).
+Example:)
+
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+pure double fun(double x) { return x * 42; }
+pure int gun(int x) { return x ^ 42; }
+pure string hun(string s) { return s; }
+auto a = fun(1.0) + fun(1.0); // can memoize
+assert(-0.0 == 0.0 && -0.0 !is 0.0);
+auto b = fun(-0.0) + fun(0.0); // can NOT memoize, -0.0 and 0.0 have different bit patterns
+auto c = hun("ab") ~ hun("ab"); // can memoize ONLY if the two strings are interned
+---
+)
+
+$(P An implementation may not evaluate a strongly pure function speculatively. This is because a
+strongly pure function may abort the program, throw, or never terminate. Example:)
```

Comment:
> may abort the program

This feels like fundamentally impure operation. I would expect a pure function may not abort. How could this occur?

> never terminate

This is suspect too; if a strong pure function enters an infinite loop, then that's just a plain program error; by definition its an infinite loop that spins the CPU doing nothing at all forever consuming no inputs and producing no outputs of any kind.
Traditionally the only reason for an infinite loop that's not a program error, is that it may break on some external signal, which a strongly pure function can never receive. It's literally just a program lockup. We should be able to assume a pure function will always terminate.

Reply:
@TurkeyMan this sounds like something at the pay grade of @tgehr and @WalterBright

Reply:
Which point?

Reply:
Whether pure functions should or should not abort, loop forever, throw errors etc.